File: go/periph/chips/imu/mpu9255.go
package imu

import (
	"fmt"
	"math"
	"time"

	"github.com/tuhde/Periph/go/periph/connection"
)

// MPU-9255 register map (subset; shared with MPU-9250 plus wake-on-motion
// helpers and a different WHO_AM_I value).
const (
	reg9255SmplrtDiv    = 0x19
	reg9255Config       = 0x1A
	reg9255GyroConfig   = 0x1B
	reg9255AccelConfig  = 0x1C
	reg9255AccelConfig2 = 0x1D
	reg9255LpAccelODR   = 0x1E
	reg9255WomThr       = 0x1F
	reg9255FifoEn       = 0x23
	reg9255IntPinCfg    = 0x37
	reg9255IntEnable    = 0x38
	reg9255IntStatus    = 0x3A
	reg9255AccelXoutH   = 0x3B
	reg9255TempOutH     = 0x41
	reg9255GyroXoutH    = 0x43
	reg9255MotDetectCtrl = 0x69
	reg9255UserCtrl     = 0x6A
	reg9255PwrMgmt1     = 0x6B
	reg9255PwrMgmt2     = 0x6C
	reg9255FifoCountH   = 0x72
	reg9255FifoR_W      = 0x74
	reg9255WhoAmI       = 0x75

	whoAmI9255Value = 0x73
)

// AK8963 magnetometer registers (address 0x0C) — identical hardware to
// MPU-9250's embedded AK8963, so reuse the constants already declared in
// mpu9250.go (same package) rather than redeclaring them here.

const (
	mpu9255ResetDelay        = 100 * time.Millisecond
	mpu9255GyroStartupDelay  = 35 * time.Millisecond
	mpu9255MagPowerDownDelay = 10 * time.Millisecond
)

// MPU9255Minimal is the MPU-9255 9-axis MotionTracking device driver
// (3-axis accelerometer + 3-axis gyroscope) — minimal interface. No
// magnetometer or wake-on-motion in Minimal; see MPU9255Full.
//
// Performs device reset, WHO_AM_I verification (always 0x73 for MPU-9255;
// MPU-9250 reads 0x71 at the same address), and configures sensible
// defaults at construction.
type MPU9255Minimal struct {
	connection connection.Connection
	accelFs    uint8
	gyroFs     uint8
}

// NewMPU9255Minimal creates a new MPU9255Minimal, resets the device,
// verifies WHO_AM_I, and configures defaults at construction.
//
// connection must be a configured I²C connection bound to the device's
// 7-bit address (0x68 default, 0x69 alternate).
func NewMPU9255Minimal(t connection.Connection) (*MPU9255Minimal, error) {
	d := &MPU9255Minimal{connection: t}
	if err := d.writeReg(reg9255PwrMgmt1, 0x80); err != nil {
		return nil, err
	}
	time.Sleep(mpu9255ResetDelay)
	if err := d.writeReg(reg9255PwrMgmt1, 0x01); err != nil {
		return nil, err
	}
	who, err := d.readReg(reg9255WhoAmI)
	if err != nil {
		return nil, err
	}
	if who != whoAmI9255Value {
		return nil, fmt.Errorf("MPU9255 WHO_AM_I: expected 0x%02X, got 0x%02X", whoAmI9255Value, who)
	}
	if err := d.writeReg(reg9255GyroConfig, 0x00); err != nil {
		return nil, err
	}
	if err := d.writeReg(reg9255AccelConfig, 0x00); err != nil {
		return nil, err
	}
	if err := d.writeReg(reg9255AccelConfig2, 0x00); err != nil {
		return nil, err
	}
	if err := d.writeReg(reg9255Config, 0x03); err != nil {
		return nil, err
	}
	if err := d.writeReg(reg9255SmplrtDiv, 0x04); err != nil {
		return nil, err
	}
	time.Sleep(mpu9255GyroStartupDelay)
	return d, nil
}

func (d *MPU9255Minimal) writeReg(reg, value byte) error {
	return d.connection.Write([]byte{reg, value})
}

func (d *MPU9255Minimal) readReg(reg byte) (byte, error) {
	b, err := d.connection.WriteRead([]byte{reg}, 1)
	if err != nil {
		return 0, err
	}
	return b[0], nil
}

func (d *MPU9255Minimal) readReg16Signed(reg byte) (int16, error) {
	b, err := d.connection.WriteRead([]byte{reg}, 2)
	if err != nil {
		return 0, err
	}
	return int16(uint16(b[0])<<8 | uint16(b[1])), nil
}

// Accel reads the 3-axis linear acceleration.
//
// Returns (x, y, z) in m/s².
func (d *MPU9255Minimal) Accel() (float32, float32, float32, error) {
	b, err := d.connection.WriteRead([]byte{reg9255AccelXoutH}, 6)
	if err != nil {
		return 0, 0, 0, err
	}
	ax := int16(uint16(b[0])<<8 | uint16(b[1]))
	ay := int16(uint16(b[2])<<8 | uint16(b[3]))
	az := int16(uint16(b[4])<<8 | uint16(b[5]))
	sens := accelSensitivity[d.accelFs]
	return float32(ax) / sens * 9.80665,
		float32(ay) / sens * 9.80665,
		float32(az) / sens * 9.80665,
		nil
}

// Gyro reads the 3-axis angular rate.
//
// Returns (x, y, z) in rad/s.
func (d *MPU9255Minimal) Gyro() (float32, float32, float32, error) {
	b, err := d.connection.WriteRead([]byte{reg9255GyroXoutH}, 6)
	if err != nil {
		return 0, 0, 0, err
	}
	gx := int16(uint16(b[0])<<8 | uint16(b[1]))
	gy := int16(uint16(b[2])<<8 | uint16(b[3]))
	gz := int16(uint16(b[4])<<8 | uint16(b[5]))
	sens := gyroSensitivity[d.gyroFs]
	const piOver180 = float32(3.141592653589793) / 180.0
	return float32(gx) / sens * piOver180,
		float32(gy) / sens * piOver180,
		float32(gz) / sens * piOver180,
		nil
}

// MPU9255Full is the MPU-9255 9-axis MotionTracking device driver — full
// interface. Extends MPU9255Minimal with configuration, magnetometer access
// via the on-board AK8963, wake-on-motion, FIFO, and power control.
//
// The magnetometer is accessed via the chip's I²C bypass mode (BYPASS_EN
// in INT_PIN_CFG); the host can then address the AK8963 directly at 0x0C.
// NewMPU9255Full takes a connection factory that produces a new I²C
// connection bound to a caller-supplied 7-bit address on the same bus.
type MPU9255Full struct {
	*MPU9255Minimal
	magConnection connection.Connection
	magEnabled    bool
	magScaleX     float32
	magScaleY     float32
	magScaleZ     float32
	magBits       uint8
}

// NewMPU9255Full creates a new MPU9255Full with the same initialization as
// NewMPU9255Minimal. The magFactory is used to construct the AK8963
// connection on demand; it must return a connection bound to the given
// 7-bit address (0x0C for the AK8963).
func NewMPU9255Full(t connection.Connection, magFactory ConnectionFactory) (*MPU9255Full, error) {
	m, err := NewMPU9255Minimal(t)
	if err != nil {
		return nil, err
	}
	if magFactory == nil {
		return nil, fmt.Errorf("MPU9255Full: magFactory is required for AK8963 access")
	}
	mag, err := magFactory(0x0C)
	if err != nil {
		return nil, fmt.Errorf("MPU9255Full: open AK8963 connection: %w", err)
	}
	return &MPU9255Full{
		MPU9255Minimal: m,
		magConnection:  mag,
	}, nil
}

// Close releases the AK8963 secondary connection. The primary MPU9255
// connection is owned by the caller and not closed here.
func (d *MPU9255Full) Close() error {
	if d.magConnection != nil {
		return d.magConnection.Close()
	}
	return nil
}

// EnableMag initialises the on-board AK8963 magnetometer and switches it
// into continuous measurement mode.
//
// bits: 14 (BIT=0) or 16 (BIT=1).
// mode: 1=single, 2=8 Hz continuous, 6=100 Hz continuous.
func (d *MPU9255Full) EnableMag(bits, mode uint8) error {
	// 1) Enable I2C bypass so the host can address the AK8963 directly.
	if err := d.writeReg(reg9255IntPinCfg, 0x22); err != nil {
		return err
	}
	// 2) Power down the AK8963, wait 10 ms.
	if err := d.magConnection.Write([]byte{ak8963CNTL1, ak8963ModePowerDown}); err != nil {
		return err
	}
	time.Sleep(mpu9255MagPowerDownDelay)
	// 3) Enter fuse ROM access mode, wait 10 ms.
	if err := d.magConnection.Write([]byte{ak8963CNTL1, ak8963ModeFuseROM}); err != nil {
		return err
	}
	time.Sleep(mpu9255MagPowerDownDelay)
	// 4) Read ASAX/Y/Z factory calibration.
	asax, err := d.magConnection.WriteRead([]byte{ak8963ASAX}, 3)
	if err != nil {
		return err
	}
	d.magScaleX = (float32(asax[0]) - 128.0) / 256.0 + 1.0
	d.magScaleY = (float32(asax[1]) - 128.0) / 256.0 + 1.0
	d.magScaleZ = (float32(asax[2]) - 128.0) / 256.0 + 1.0
	// 5) Power down, wait 10 ms.
	if err := d.magConnection.Write([]byte{ak8963CNTL1, ak8963ModePowerDown}); err != nil {
		return err
	}
	time.Sleep(mpu9255MagPowerDownDelay)
	// 6) Configure continuous mode at the requested resolution.
	var bitFlag uint8
	if bits == 16 {
		bitFlag = 0x10
	}
	if err := d.magConnection.Write([]byte{ak8963CNTL1, bitFlag | (mode & 0x0F)}); err != nil {
		return err
	}
	d.magEnabled = true
	d.magBits = bits
	return nil
}

// Mag reads the 3-axis magnetic field strength.
//
// Returns (x, y, z) in µT. Mag must have been enabled first via EnableMag.
func (d *MPU9255Full) Mag() (float32, float32, float32, error) {
	if !d.magEnabled {
		return 0, 0, 0, fmt.Errorf("MPU9255 Mag: magnetometer not enabled (call EnableMag first)")
	}
	// Read 7 bytes from 0x03 (HXL) through 0x09 (ST2). ST2 must be read
	// to unlock the next measurement.
	b, err := d.magConnection.WriteRead([]byte{ak8963HXL}, 7)
	if err != nil {
		return 0, 0, 0, err
	}
	if b[6]&0x08 != 0 {
		return 0, 0, 0, fmt.Errorf("MPU9255 Mag: AK8963 overflow")
	}
	mx := int16(uint16(b[1])<<8 | uint16(b[0]))
	my := int16(uint16(b[3])<<8 | uint16(b[2]))
	mz := int16(uint16(b[5])<<8 | uint16(b[4]))
	var sens float32
	if d.magBits == 16 {
		sens = 0.15
	} else {
		sens = 0.6
	}
	return float32(mx) * sens * d.magScaleX,
		float32(my) * sens * d.magScaleY,
		float32(mz) * sens * d.magScaleZ,
		nil
}

// ConfigureGyro sets the gyroscope full-scale range.
func (d *MPU9255Full) ConfigureGyro(fullScale uint8) error {
	d.gyroFs = fullScale & 0x03
	return d.writeReg(reg9255GyroConfig, (fullScale&0x03)<<3)
}

// ConfigureAccel sets the accelerometer full-scale range.
func (d *MPU9255Full) ConfigureAccel(fullScale uint8) error {
	d.accelFs = fullScale & 0x03
	return d.writeReg(reg9255AccelConfig, (fullScale&0x03)<<3)
}

// ConfigureDLPF sets the gyroscope and accelerometer DLPF bandwidths.
func (d *MPU9255Full) ConfigureDLPF(gyroDLPF, accelDLPF uint8) error {
	if err := d.writeReg(reg9255Config, gyroDLPF&0x07); err != nil {
		return err
	}
	return d.writeReg(reg9255AccelConfig2, accelDLPF&0x07)
}

// ConfigureSampleRate sets the sample rate divider.
func (d *MPU9255Full) ConfigureSampleRate(divider uint8) error {
	return d.writeReg(reg9255SmplrtDiv, divider)
}

// Temperature reads the die temperature in °C.
func (d *MPU9255Full) Temperature() (float32, error) {
	raw, err := d.readReg16Signed(reg9255TempOutH)
	if err != nil {
		return 0, err
	}
	return float32(raw)/333.87 + 21.0, nil
}

// AccelRaw reads the raw 3-axis accelerometer values.
func (d *MPU9255Full) AccelRaw() (int16, int16, int16, error) {
	b, err := d.connection.WriteRead([]byte{reg9255AccelXoutH}, 6)
	if err != nil {
		return 0, 0, 0, err
	}
	return int16(uint16(b[0])<<8 | uint16(b[1])),
		int16(uint16(b[2])<<8 | uint16(b[3])),
		int16(uint16(b[4])<<8 | uint16(b[5])),
		nil
}

// GyroRaw reads the raw 3-axis gyroscope values.
func (d *MPU9255Full) GyroRaw() (int16, int16, int16, error) {
	b, err := d.connection.WriteRead([]byte{reg9255GyroXoutH}, 6)
	if err != nil {
		return 0, 0, 0, err
	}
	return int16(uint16(b[0])<<8 | uint16(b[1])),
		int16(uint16(b[2])<<8 | uint16(b[3])),
		int16(uint16(b[4])<<8 | uint16(b[5])),
		nil
}

// MagRaw reads the raw magnetometer values.
func (d *MPU9255Full) MagRaw() (int16, int16, int16, error) {
	if !d.magEnabled {
		return 0, 0, 0, fmt.Errorf("MPU9255 MagRaw: magnetometer not enabled (call EnableMag first)")
	}
	b, err := d.magConnection.WriteRead([]byte{ak8963HXL}, 7)
	if err != nil {
		return 0, 0, 0, err
	}
	return int16(uint16(b[1])<<8 | uint16(b[0])),
		int16(uint16(b[3])<<8 | uint16(b[2])),
		int16(uint16(b[5])<<8 | uint16(b[4])),
		nil
}

// DataReady returns true when RAW_DATA_RDY_INT is set in INT_STATUS.
func (d *MPU9255Full) DataReady() (bool, error) {
	s, err := d.readReg(reg9255IntStatus)
	if err != nil {
		return false, err
	}
	return s&0x01 != 0, nil
}

// SetSleep sets or clears the SLEEP bit in PWR_MGMT_1.
func (d *MPU9255Full) SetSleep(sleep bool) error {
	v, err := d.readReg(reg9255PwrMgmt1)
	if err != nil {
		return err
	}
	if sleep {
		v |= 0x40
	} else {
		v &^= 0x40
	}
	return d.writeReg(reg9255PwrMgmt1, v)
}

// FIFOcount returns the number of bytes in the FIFO.
func (d *MPU9255Full) FIFOcount() (uint16, error) {
	b, err := d.connection.WriteRead([]byte{reg9255FifoCountH}, 2)
	if err != nil {
		return 0, err
	}
	return (uint16(b[0])&0x1F)<<8 | uint16(b[1]), nil
}

// ReadFIFO reads up to len(buf) bytes from the FIFO.
func (d *MPU9255Full) ReadFIFO(buf []byte) (uint16, error) {
	count, err := d.FIFOcount()
	if err != nil {
		return 0, err
	}
	if count == 0 {
		return 0, nil
	}
	toRead := int(count)
	if toRead > len(buf) {
		toRead = len(buf)
	}
	read, err := d.connection.WriteRead([]byte{reg9255FifoR_W}, toRead)
	if err != nil {
		return 0, err
	}
	n := copy(buf, read)
	return uint16(n), nil
}

// EnableFIFO configures and enables FIFO sources.
func (d *MPU9255Full) EnableFIFO(gyro, accel, temp bool) error {
	var v uint8
	if accel {
		v |= 1 << 3
	}
	if temp {
		v |= 1 << 2
	}
	if gyro {
		v |= 1 << 4
	}
	if err := d.writeReg(reg9255FifoEn, v); err != nil {
		return err
	}
	uc, err := d.readReg(reg9255UserCtrl)
	if err != nil {
		return err
	}
	return d.writeReg(reg9255UserCtrl, uc|0x40)
}

// ResetFIFO resets the FIFO buffer.
func (d *MPU9255Full) ResetFIFO() error {
	uc, err := d.readReg(reg9255UserCtrl)
	if err != nil {
		return err
	}
	return d.writeReg(reg9255UserCtrl, uc|0x04)
}

// lposcTable maps LP_ACCEL_ODR's Lposc_clksel field to the chip's published
// wake-up output data rate (Hz). Used by ConfigureWakeOnMotion.
var lposcTable = [16]float32{
	0.24, 0.49, 0.98, 1.95, 3.91, 7.81, 15.63, 31.25,
	62.5, 125.0, 250.0, 500.0, 1000.0, 2000.0, 4000.0, 8000.0,
}

// ConfigureWakeOnMotion arms the hardware motion-detection logic and
// enters accelerometer-only low-power mode. Disables the gyroscope, sets
// the accelerometer DLPF to ~184 Hz, arms WOM_THR / MOT_DETECT_CTRL, and
// enables CYCLE mode in PWR_MGMT_1.
//
// thresholdMg: motion threshold in milligrams (4–1020 mg, quantized to
//              4 mg steps; values outside the range are clamped).
// odrHz:       wake-up output data rate in Hz; mapped to the chip's
//              standard 16-entry LP_ACCEL_ODR table (closest match).
func (d *MPU9255Full) ConfigureWakeOnMotion(thresholdMg uint16, odrHz float32) error {
	thresholdLsb := int((uint32(thresholdMg) + 2) / 4)
	if thresholdLsb < 1 {
		thresholdLsb = 1
	}
	if thresholdLsb > 255 {
		thresholdLsb = 255
	}
	bestSel := 0
	bestDiff := math.Abs(float64(odrHz - lposcTable[0]))
	for sel := 1; sel < len(lposcTable); sel++ {
		diff := math.Abs(float64(odrHz - lposcTable[sel]))
		if diff < bestDiff {
			bestDiff = diff
			bestSel = sel
		}
	}

	if err := d.writeReg(reg9255PwrMgmt1, 0x01); err != nil {
		return err
	}
	if err := d.writeReg(reg9255PwrMgmt2, 0x07); err != nil {
		return err
	}
	if err := d.writeReg(reg9255AccelConfig2, 0x09); err != nil {
		return err
	}
	if err := d.writeReg(reg9255IntEnable, 0x40); err != nil {
		return err
	}
	if err := d.writeReg(reg9255MotDetectCtrl, 0xC0); err != nil {
		return err
	}
	if err := d.writeReg(reg9255WomThr, byte(thresholdLsb)); err != nil {
		return err
	}
	if err := d.writeReg(reg9255LpAccelODR, byte(bestSel&0x0F)); err != nil {
		return err
	}
	return d.writeReg(reg9255PwrMgmt1, 0x21)
}

// MotionDetected returns true when WOM_INT (bit 6) is set in INT_STATUS.
// Reading INT_STATUS clears the interrupt.
func (d *MPU9255Full) MotionDetected() (bool, error) {
	s, err := d.readReg(reg9255IntStatus)
	if err != nil {
		return false, err
	}
	return s&0x40 != 0, nil
}

File: cpp/src/chips/imu/MPU9255.cpp
#include "MPU9255.h"
#include <stdlib.h>

#if defined(__linux__)
#include <unistd.h>
#define DELAY_MS(ms) usleep((ms) * 1000)
#elif defined(__ZEPHYR__)
#include <zephyr/kernel.h>
#define DELAY_MS(ms) k_sleep(K_MSEC(ms))
#elif defined(ESP_PLATFORM)
#include <freertos/FreeRTOS.h>
#include <freertos/task.h>
#define DELAY_MS(ms) vTaskDelay(pdMS_TO_TICKS(ms))
#elif __has_include(<pico/time.h>)
#include <pico/time.h>
#define DELAY_MS(ms) sleep_ms(ms)
#else
#include <Arduino.h>
#define DELAY_MS(ms) delay(ms)
#endif

MPU9255Minimal::MPU9255Minimal(Connection& connection)
    : _connection(connection) {
    _write_reg(REG_PWR_MGMT_1, 0x80);
    DELAY_MS(100);
    _write_reg(REG_PWR_MGMT_1, 0x01);
    uint8_t who = _read_reg(REG_WHO_AM_I);
    if (who != WHO_AM_I_VALUE) {
        abort();
    }
    _write_reg(REG_GYRO_CONFIG, 0x00);
    _write_reg(REG_ACCEL_CONFIG, 0x00);
    _write_reg(REG_ACCEL_CONFIG2, 0x03);
    _write_reg(REG_CONFIG, 0x03);
    _write_reg(REG_SMPLRT_DIV, 0x04);
    DELAY_MS(35);
}

void MPU9255Minimal::_write_reg(uint8_t reg, uint8_t value) {
    uint8_t buf[2] = { reg, value };
    _connection.write(buf, 2);
}

uint8_t MPU9255Minimal::_read_reg(uint8_t reg) {
    uint8_t val;
    _connection.write_read(&reg, 1, &val, 1);
    return val;
}

int16_t MPU9255Minimal::_read_reg16_signed(uint8_t reg) {
    uint8_t buf[2];
    _connection.write_read(&reg, 1, buf, 2);
    return static_cast<int16_t>((static_cast<uint16_t>(buf[0]) << 8) | buf[1]);
}

void MPU9255Minimal::_read_burst(uint8_t reg, uint8_t* buf, uint8_t len) {
    _connection.write_read(&reg, 1, buf, len);
}

void MPU9255Minimal::accel(float& x, float& y, float& z) {
    uint8_t buf[6];
    _read_burst(REG_ACCEL_XOUT_H, buf, 6);
    int16_t ax = static_cast<int16_t>((static_cast<uint16_t>(buf[0]) << 8) | buf[1]);
    int16_t ay = static_cast<int16_t>((static_cast<uint16_t>(buf[2]) << 8) | buf[3]);
    int16_t az = static_cast<int16_t>((static_cast<uint16_t>(buf[4]) << 8) | buf[5]);
    float sens = ACCEL_SENSITIVITY[_accel_fs];
    x = ax / sens * 9.80665f;
    y = ay / sens * 9.80665f;
    z = az / sens * 9.80665f;
}

void MPU9255Minimal::gyro(float& x, float& y, float& z) {
    uint8_t buf[6];
    _read_burst(REG_GYRO_XOUT_H, buf, 6);
    int16_t gx = static_cast<int16_t>((static_cast<uint16_t>(buf[0]) << 8) | buf[1]);
    int16_t gy = static_cast<int16_t>((static_cast<uint16_t>(buf[2]) << 8) | buf[3]);
    int16_t gz = static_cast<int16_t>((static_cast<uint16_t>(buf[4]) << 8) | buf[5]);
    float sens = GYRO_SENSITIVITY[_gyro_fs];
    x = gx / sens * 3.141592653589793f / 180.0f;
    y = gy / sens * 3.141592653589793f / 180.0f;
    z = gz / sens * 3.141592653589793f / 180.0f;
}

MPU9255Full::MPU9255Full(Connection& connection, Connection& magConnection)
    : MPU9255Minimal(connection), _mag_connection(magConnection) {}

void MPU9255Full::configure_gyro(uint8_t full_scale) {
    _gyro_fs = full_scale & 0x03;
    _write_reg(REG_GYRO_CONFIG, (full_scale & 0x03) << 3);
}

void MPU9255Full::configure_accel(uint8_t full_scale) {
    _accel_fs = full_scale & 0x03;
    _write_reg(REG_ACCEL_CONFIG, (full_scale & 0x03) << 3);
}

void MPU9255Full::configure_dlpf(uint8_t gyro_dlpf, uint8_t accel_dlpf) {
    _write_reg(REG_CONFIG, gyro_dlpf & 0x07);
    _write_reg(REG_ACCEL_CONFIG2, accel_dlpf & 0x07);
}

void MPU9255Full::configure_sample_rate(uint8_t divider) {
    _write_reg(REG_SMPLRT_DIV, divider);
}

float MPU9255Full::temperature() {
    int16_t raw = _read_reg16_signed(REG_TEMP_OUT_H);
    return raw / 333.87f + 21.0f;
}

void MPU9255Full::_ak8963_write(uint8_t reg, uint8_t value) {
    uint8_t buf[2] = { reg, value };
    _mag_connection.write(buf, 2);
}

uint8_t MPU9255Full::_ak8963_read(uint8_t reg) {
    uint8_t val;
    _mag_connection.write_read(&reg, 1, &val, 1);
    return val;
}

void MPU9255Full::_ak8963_read_burst(uint8_t reg, uint8_t* buf, uint8_t len) {
    _mag_connection.write_read(&reg, 1, buf, len);
}

void MPU9255Full::enable_mag(uint8_t bits, uint8_t mode) {
    _write_reg(REG_INT_PIN_CFG, 0x22);
    DELAY_MS(10);

    _ak8963_write(AK8963_REG_CNTL1, 0x00);
    DELAY_MS(10);

    _ak8963_write(AK8963_REG_CNTL1, 0x0F);
    DELAY_MS(10);

    uint8_t asax = _ak8963_read(AK8963_REG_ASAX);
    uint8_t asay = _ak8963_read(AK8963_REG_ASAY);
    uint8_t asaz = _ak8963_read(AK8963_REG_ASAZ);

    _mag_scale_x = (static_cast<float>(asax) - 128.0f) / 256.0f + 1.0f;
    _mag_scale_y = (static_cast<float>(asay) - 128.0f) / 256.0f + 1.0f;
    _mag_scale_z = (static_cast<float>(asaz) - 128.0f) / 256.0f + 1.0f;

    _ak8963_write(AK8963_REG_CNTL1, 0x00);
    DELAY_MS(10);

    uint8_t cntl1_val = 0;
    if (bits == 16) {
        cntl1_val |= 0x10;
    }
    cntl1_val |= (mode & 0x0F);
    _ak8963_write(AK8963_REG_CNTL1, cntl1_val);
    DELAY_MS(10);

    _mag_enabled = true;
    _mag_bits = bits;
}

void MPU9255Full::mag(float& x, float& y, float& z) {
    if (!_mag_enabled) {
        return;
    }
    uint8_t buf[7];
    _ak8963_read_burst(AK8963_REG_HXL, buf, 7);
    int16_t mx = static_cast<int16_t>((static_cast<uint16_t>(buf[1]) << 8) | buf[0]);
    int16_t my = static_cast<int16_t>((static_cast<uint16_t>(buf[3]) << 8) | buf[2]);
    int16_t mz = static_cast<int16_t>((static_cast<uint16_t>(buf[5]) << 8) | buf[4]);
    uint8_t st2 = buf[6];
    (void)st2;

    float sens = (_mag_bits == 16) ? MAG_SENSITIVITY_16BIT : MAG_SENSITIVITY_14BIT;
    x = static_cast<float>(mx) * sens * _mag_scale_x;
    y = static_cast<float>(my) * sens * _mag_scale_y;
    z = static_cast<float>(mz) * sens * _mag_scale_z;
}

void MPU9255Full::accel_raw(int16_t& x, int16_t& y, int16_t& z) {
    uint8_t buf[6];
    _read_burst(REG_ACCEL_XOUT_H, buf, 6);
    x = static_cast<int16_t>((static_cast<uint16_t>(buf[0]) << 8) | buf[1]);
    y = static_cast<int16_t>((static_cast<uint16_t>(buf[2]) << 8) | buf[3]);
    z = static_cast<int16_t>((static_cast<uint16_t>(buf[4]) << 8) | buf[5]);
}

void MPU9255Full::gyro_raw(int16_t& x, int16_t& y, int16_t& z) {
    uint8_t buf[6];
    _read_burst(REG_GYRO_XOUT_H, buf, 6);
    x = static_cast<int16_t>((static_cast<uint16_t>(buf[0]) << 8) | buf[1]);
    y = static_cast<int16_t>((static_cast<uint16_t>(buf[2]) << 8) | buf[3]);
    z = static_cast<int16_t>((static_cast<uint16_t>(buf[4]) << 8) | buf[5]);
}

void MPU9255Full::mag_raw(int16_t& x, int16_t& y, int16_t& z) {
    if (!_mag_enabled) {
        x = y = z = 0;
        return;
    }
    // ST2 (buf[6]) is not used but must be read to unlock the next measurement.
    uint8_t buf[7];
    _ak8963_read_burst(AK8963_REG_HXL, buf, 7);
    x = static_cast<int16_t>((static_cast<uint16_t>(buf[1]) << 8) | buf[0]);
    y = static_cast<int16_t>((static_cast<uint16_t>(buf[3]) << 8) | buf[2]);
    z = static_cast<int16_t>((static_cast<uint16_t>(buf[5]) << 8) | buf[4]);
}

bool MPU9255Full::data_ready() {
    return (_read_reg(REG_INT_STATUS) & 0x01) != 0;
}

void MPU9255Full::set_sleep(bool sleep) {
    uint8_t val = _read_reg(REG_PWR_MGMT_1);
    if (sleep) {
        val |= 0x40;
    } else {
        val &= ~0x40;
    }
    _write_reg(REG_PWR_MGMT_1, val);
}

uint16_t MPU9255Full::fifo_count() {
    uint8_t buf[2];
    _read_burst(REG_FIFO_COUNTH, buf, 2);
    return ((static_cast<uint16_t>(buf[0]) & 0x1F) << 8) | buf[1];
}

uint16_t MPU9255Full::read_fifo(uint8_t* buf, uint16_t len) {
    uint16_t count = fifo_count();
    if (count == 0) return 0;
    uint16_t to_read = (count < len) ? count : len;
    _read_burst(REG_FIFO_R_W, buf, to_read);
    return to_read;
}

void MPU9255Full::enable_fifo(bool gyro, bool accel, bool temp) {
    uint8_t fifo_en = ((accel ? 1 : 0) << 3) | ((temp ? 1 : 0) << 2) | ((gyro ? 1 : 0) << 4);
    _write_reg(REG_FIFO_EN, fifo_en);
    uint8_t user_ctrl = _read_reg(REG_USER_CTRL);
    _write_reg(REG_USER_CTRL, user_ctrl | 0x40);
}

void MPU9255Full::reset_fifo() {
    uint8_t user_ctrl = _read_reg(REG_USER_CTRL);
    _write_reg(REG_USER_CTRL, user_ctrl | 0x04);
}

void MPU9255Full::configure_wake_on_motion(uint16_t threshold_mg, float odr_hz) {
    int threshold_lsb = (static_cast<int>(threshold_mg) + 2) / 4;
    if (threshold_lsb < 1) threshold_lsb = 1;
    if (threshold_lsb > 255) threshold_lsb = 255;

    static const float lposc_table[16] = {
        0.24f, 0.49f, 0.98f, 1.95f, 3.91f, 7.81f, 15.63f, 31.25f,
        62.5f, 125.0f, 250.0f, 500.0f, 1000.0f, 2000.0f, 4000.0f, 8000.0f,
    };
    int best_sel = 0;
    float best_diff = (odr_hz < lposc_table[0]) ? (lposc_table[0] - odr_hz) : (odr_hz - lposc_table[0]);
    for (int sel = 1; sel < 16; ++sel) {
        float diff = (odr_hz < lposc_table[sel]) ? (lposc_table[sel] - odr_hz) : (odr_hz - lposc_table[sel]);
        if (diff < best_diff) {
            best_diff = diff;
            best_sel = sel;
        }
    }

    _write_reg(REG_PWR_MGMT_1, 0x01);
    _write_reg(REG_PWR_MGMT_2, 0x07);
    _write_reg(REG_ACCEL_CONFIG2, 0x01);
    _write_reg(REG_INT_ENABLE, 0x40);
    _write_reg(REG_MOT_DETECT_CTRL, 0xC0);
    _write_reg(REG_WOM_THR, static_cast<uint8_t>(threshold_lsb));
    _write_reg(REG_LP_ACCEL_ODR, static_cast<uint8_t>(best_sel & 0x0F));
    _write_reg(REG_PWR_MGMT_1, 0x21);
}

bool MPU9255Full::motion_detected() {
    return (_read_reg(REG_INT_STATUS) & 0x40) != 0;
}